fuse/api.go

```diff
@@ -2,9 +2,74 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// The fuse package provides APIs to implement filesystems in
-// userspace.  Typically, each call of the API happens in its own
+// Package fuse provides APIs to implement filesystems in
+// userspace in terms of raw FUSE protocol.
+//
+// A filesystem is implemented by implementing its server that provides a
+// RawFileSystem interface. Typically the server embeds
+// NewDefaultRawFileSystem() and implements only subset of filesystem methods:
+//
+//	type MyFS struct {
+//		fuse.RawFileSystem
+//		...
+//	}
+//
+//	func MyFSNew() *MyFS {
+//		return &MyFS{
+//			RawFileSystem: fuse.NewDefaultRawFileSystem(),
+//			...
+//		}
+//	}
+//
+//	// Mkdir implements "mkdir" request handler.
+//	//
+//	// For other requests - not explicitly implemented by MyFS - ENOSYS
+//	// will be typically returned to client.
+//	func (fs *MyFS) Mkdir(...) {
+//		...
+//	}
+//
+// Then the filesystem can be mounted and served to a client (typically OS
+// kernel) by creating Server:
+//
+//	fs := MyFSNew() // implements RawFileSystem
+//	fssrv, err := fuse.NewServer(fs, mountpoint, &fuse.MountOptions{...})
+//	if err != nil {
+//		...
+//	}
+//
+//	fssrv.Serve()
+//
```

Comment:
there should be a call for WaitMount() somewhere, if the server wants to interact with the FS too, like in tests.

Reply:
Agree. I did not put it in there becuase my filesystem runs .Serve() synchronously, but you are right. Reworking this to:

```diff
 //     if err != nil {
 //             ...
 //     }
 //
+// and letting the server do its work:
+//
+//     // either synchronously - .Serve() blocks until the filesystem is unmounted.
 //     fssrv.Serve()
 //
+//     // or in the background - .Serve() is spawned in another goroutine, but
+//     // before interacting with fssrv from current context we have to wait
+//     // until the filesystem mounting is complete.
+//     go fssrv.Serve()
+//     err = fssrv.WaitMount()
+//     if err != nil {
+//             ...
+//     }
+//
```

By the way - `.Serve()` does not return any error - is that good? In my view `.Serve()` should always return an error, and the error returned should indicate the reason Serve finished - either e.g. filesystem was unmounted normally, or another e.g. IO error with the kernel via /dev/fuse or whatever else.

Maybe we can add the error return to `.Serve()` because it does not break compatibility?